File: examples/python/ds/api/book.py
import ds.base.book as _base
from ds.base._core import Submanager
from ds.model._reference import *


class Book(_base.Book):

    class Publisher(Submanager):

        def get(self):
            return None if (self._.publisher_id is None) else self.api.publishers.one(self._.publisher_id)

        def set(self, x:_base.PublisherRef):
            _base.PublisherRef.valid(x)
            self._.publisher_id = int(x)
            self.api.books.set(x)

    class Genres(Submanager):

        def add(self, x, rank=1):
            return self.api.books.add(BookGenreRef(self._, int(x)))

        def remove(self, x):
            return self.api.books.remove(BookGenreRef(self._, int(x)))

        def all(self):
            return self.api.authors.all(BookRef(self._))

    class Authors(Submanager):

        def add(self, x, rank=1):
            return self.api.books.add(BookAuthor(self._, int(x), rank))

        def remove(self, x):
            return self.api.books.remove(BookAuthorRef(self._, int(x)))

        def all(self):
            return self.api.genres.all(BookRef(self._))

    def __init__(self, publisher_id=None, title=None, year=None, id=None, api=None):
        super().__init__(publisher_id, title, year, id, api)
        self._publisher = Book.Publisher(self)
        self._genres = Book.Genres(self)
        self._authors = Book.Authors(self)

    @property
    def publisher(self):
        return self._publisher

    @property
    def genres(self):
        return self._genres

    @property
    def authors(self):
        return self._authors


class BookManager(_base.BookManager):

    def _create(self, x):
        return Book(id=x[0], publisher_id=x[1], title=x[2], year=x[3]).attach(self.api)

File: examples/php/ds/api/book.php
<?php
namespace books;
require_once __DIR__ . '/../data/book.php';

// Foreign: 1

class PublisherSubset {
    private model\Book $book;

    function __construct(model\Book $book) {
        $this->book = $book;
    }

    function get() {
        return is_null($this->book->publisher_id) ? null : $this->book->api->publishers->one($this->_->publisher_id);
    }

    function set(PublisherRef $x) {
        PublisherRef::valid($x);
        $this->book->publisher_id = $x->id;
        $this->api->books->set($x);
    }
}


class Book extends model\Book {
    public $api = null;
    private PublisherSubset $publisher;

    function __construct($publisher_id=null, $title=null, $year=null, $id=null, $api=null) {
        parent::__construct($publisher_id, $title, $year, $id);
        $this->api = $api;
        $this->publisher = new PublisherSubset($this);
    }

    function add($x) {
        if ($x instanceof Engine) {
            return $x->books->add($this);
        }
    }

    function remove() {
        $this->api->books->remove($this);
        $this->id = null;
    }

    function push(?Fields $x=null) {
        return $this->api->books->set(is_null($x) ? $this : $x);
    }

    function pull(?Fields $x=null) {
        return $this->api->books->get($x);
    }
}


class BookManager extends BookData {
    private $api;
    
    function __construct(\quartz\Database $db, $api) {
        parent::__construct($db);
        $this->api = $api;
    }

    function fetch($x, ?int $id=null, ?bool $debug=false): Book {
        $x_ = strval($x);
        $b = $this->one($x_);
        if (is_null($b)) {
            $b = new Book(id:$id, title:$x_, api:$this->api);
            $this->add($b, $debug);
        }
        return $b;
    }

    protected function create($x) {
        return new Book(id:$x[0], publisher_id:$x[1], title:$x[2], year:$x[3], api:$this->api);
    }
}

?>